### PR TITLE
IADCSensorData 

### PR DESCRIPTION
**Pull request by Andromas14**
# Purpose

Create IADCSensorData class overseeing the connection and processing of the data coming from most sensors

## Implementation

Gets data from the ADCs and returns processed data. All sensors need to be declared as single objects, and the Sensor.Process() function will be run in the main loop, to gather data from each sensor.

## Testing 

It works with an ESP32, but needs further testing with ADCs and sensors.

View on github

## References

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#referencing-issues-and-pull-requests

## PR Review Rules 

1. Reviewer is responsible for resolving any threads they started
2. One commit per thread and if it's a widespread change, link the commit in the reply pls :)
3. ...

### CODE REVIEW

File: DAQ_FW/lib/IADCSensor/IADCSensor.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    // obj declared when logger not active, could be hard to log
+    Logger::Notice("Class created with gain mode: %u and V/per bit of %D", _ADC_GAIN_MODE, mVOLTAGE_PER_BIT);
+}
+
+float IADCSensor::Process()
+{
+
+    // TODO: put Read() function here once ready, currently for testing
+    uint16_t raw_data = 5230;
```
Comment (rafguevara14) [inline DAQ_FW/lib/IADCSensor/IADCSensor.cpp]: We should be able to call Read() here it just won't do anything until @lonamle overrides the function.

File: DAQ_FW/lib/IADCSensor/IADCSensor.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    uint32_t mSensorID;
+    uint32_t mADC_ID;
+
+    void ADCcovert(uint16_t _ADC_GAIN_MODE);
+
+public:
+    IADCSensor(const char *_SensorName, uint16_t _SensorID, uint16_t _ADC_ID, uint16_t _ADC_GAIN_MODE);
+
+    // Read function - to be overwritten by the Driver class
+    virtual void Read(){};
```
Comment (rafguevara14) [inline DAQ_FW/lib/IADCSensor/IADCSensor.h]: This should either be abstract or have a default implementation to catch the mistake of a child not implementing this function. The former gives us an error at compilation and the latter logs the mistake at runtime. 

Abstract: `virtual void Read() = 0;` this means no child can be defined _unless_ an implementation for the `Read()` function is provided. This enforces every child to implement it

Default: If the function is not overridden the default implementation is used that could return an error or log the error to the console.

I think making the function abstract is the right call here I think since **every** sensor needs a `Read()` function to

File: DAQ_FW/lib/IADCSensor/IADCSensor.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    float final_data = raw_data * mVOLTAGE_PER_BIT;
+
+    Logger::Notice("Processed Data: %D, from sensor <%s> with ID: %u and adc id %u", final_data, mSensorName, mSensorID, mADC_ID);
+
+    return (float)final_data;
+}
+
+// picks correct Voltage ber bit ration, according to the gain mode
+void IADCSensor::ADCcovert(uint16_t _ADC_GAIN_MODE)
```
Comment (rafguevara14) [inline DAQ_FW/lib/IADCSensor/IADCSensor.cpp]: I didn't realize this function was initializing `mVOLTAGE_PER_BIT` which I think is for two reasons: the naming of the function is a little misleading and proper `const correctness` which I'll mention in another comment.

If you want to keep this function instead of moving the body to the constructor, what it's really doing is not converting the ADC, but initializing the voltage per bit. `initializeVoltagePerBit()` or something similar would be better

File: DAQ_FW/lib/IADCSensor/IADCSensor.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+IADCSensor::IADCSensor(const char *_SensorName, uint16_t _SensorID, uint16_t _ADC_ID, uint16_t _ADC_GAIN_MODE)
+    : mSensorName(_SensorName), mSensorID(_SensorID), mADC_ID(_ADC_ID)
+{
+    ADCcovert(_ADC_GAIN_MODE);
+
+    // obj declared when logger not active, could be hard to log
+    Logger::Notice("Class created with gain mode: %u and V/per bit of %D", _ADC_GAIN_MODE, mVOLTAGE_PER_BIT);
+}
+
+float IADCSensor::Process()
```
Comment (rafguevara14) [inline DAQ_FW/lib/IADCSensor/IADCSensor.cpp]: `Const correctness` is just a C++ habit of labeling methods as const if they don't modify the member variables. This makes it easier to debug when you find that member variables are changing unexpectedly; instead of checking every method, you can ignore ones labeled as const.

As good practice, you should try and figure out if your method will be const when you first declare them as it gives you a stricter guideline when actually writing the implementation. If you accidentally modify a member variable you'll get a compile-time error.

More info here: https://stackoverflow.com/questions/136880/sell-me-const-correctness 

TLDR: This method I think should be const

File: DAQ_FW/lib/IADCSensor/IADCSensor.h (hunk 2 above)
Comment (rafguevara14) [inline DAQ_FW/lib/IADCSensor/IADCSensor.h]: I also think this method should be const. This one is even more important because it ensures that @lonamle doesn't accidentally modify the member variables you declared and gives her more info on how to write her function

File: DAQ_FW/lib/IADCSensor/IADCSensor.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#ifndef IADCSENSOR_LIB
+#define IADCSENSOR_LIB
+
+#include "Logger.h"
+#include "system_config.h"
+
+class IADCSensor
```
Comment (rafguevara14) [inline DAQ_FW/lib/IADCSensor/IADCSensor.h:7]: for general cleanliness, class definitions should be organized in terms of relevance to an outside user. If I'm trying to use this class, I really only care about the constructor and read functions, not the stuff I can't even access like private members. So here's a suggested order. It kinda varies depending on who's code style you follow but at the very least constructors/destructors and then public methods should be first 

```
public: 
     constructors/destructors
     methods
     members
    
private:
    members 
    methods

```

**Review by rafguevara14 — CHANGES_REQUESTED** [review]
I think this is the first PR since I added these rules so I'm just gonna reiterate here!

```
PR Review Rules
Reviewer is responsible for resolving any threads they started
One commit per thread and if it's a widespread change, link the commit in the reply pls :)
...
```

File: DAQ_FW/lib/IADCSensor/IADCSensor.h (hunk 2 above)
Comment (Andromas14) [reply DAQ_FW/lib/IADCSensor/IADCSensor.h]: Is there a way to implement the abstract version currently without getting a compiler error? My understanding is that it needs at least one derived class with the method definition to run, which we currently don't have. I can add the error definition for the default version for now, and make it abstract once the derived class is there.

Comment (rafguevara14) [reply DAQ_FW/lib/IADCSensor/IADCSensor.h]: Did you check that it doesn't compile? I believe the compiler only requires that any children must implement the method, **not** that there necessarily needs to be a derived class. If I'm wrong though you can do that for sure

Comment (Andromas14) [reply DAQ_FW/lib/IADCSensor/IADCSensor.h]: image
It gives this error, setting `virtual void Read() = 0;` makes the class abstract, thus the object from the base class cannot be declared.

Comment (rafguevara14) [reply DAQ_FW/lib/IADCSensor/IADCSensor.h]: Ohh I see what you mean. Let's just remove that line of code. You can add it as a comment in the header file of the class if you want

Comment (rafguevara14) [reply DAQ_FW/lib/IADCSensor/IADCSensor.h]: The current setup is fine. Disregard!

**Review by rafguevara14 — APPROVED** [review]
LGTM !

**Review by rafguevara14 — APPROVED** [review]
LGTM !
